go/tasks/plugins/k8s/sagemaker/sagemaker.go

```diff
@@ -194,13 +203,117 @@ func (m awsSagemakerPlugin) BuildResourceForTrainingJob(
 			},
 			TensorBoardOutputConfig: nil,
 			Tags:                    nil,
-			TrainingJobName:         &taskName,
+			TrainingJobName:         &jobName,
 		},
 	}
 	logger.Infof(ctx, "Successfully built a training job resource for task [%v]", taskCtx.TaskExecutionMetadata().GetTaskExecutionID().GetGeneratedName())
 	return trainingJob, nil
 }
 
+func (m awsSagemakerPlugin) BuildResourceForCustomTrainingJob(
+	ctx context.Context, taskCtx pluginsCore.TaskExecutionContext) (k8s.Resource, error) {
+
+	logger.Infof(ctx, "Building a training job resource for task [%v]", taskCtx.TaskExecutionMetadata().GetTaskExecutionID().GetGeneratedName())
+	taskTemplate, err := getTaskTemplate(ctx, taskCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal the custom field of the task template back into the Hyperparameter Tuning Job struct generated in flyteidl
+	sagemakerTrainingJob := flyteSageMakerIdl.TrainingJob{}
+	err = utils.UnmarshalStruct(taskTemplate.GetCustom(), &sagemakerTrainingJob)
+	if err != nil {
+		return nil, errors.Wrapf(err, "invalid TrainingJob task specification: not able to unmarshal the custom field to [%s]", m.TaskType)
+	}
+
+	inputChannels := make([]commonv1.Channel, 0)
+	inputModeString := strings.Title(strings.ToLower(sagemakerTrainingJob.GetAlgorithmSpecification().GetInputMode().String()))
```

Comment:
Just to note: @EngHabu and I had a discussion earlier and decided to add guarding statements in both custom and built-in training plugin to fail the job early if incorrect algorithm name is passed in. I also made this change in flytekit (https://github.com/lyft/flytekit/pull/156/commits/9aeef7249b860d89032a977f99808f79d94bdc8d) to default the algorithm name of an algorithm specification to Custom, so that when users write a custom training job, they do not need to specify the `algorithm_name` and the `algorithm_version` fields.

We also conclude that a future breaking change should be introduced to make the interface more reasonable for the users.